### PR TITLE
fixes dotnet-diagnostictests funceval failures

### PR DESCRIPTION
**Pull request by mikelle-rogers**
instantiates all elements of the array

### CODE REVIEW

**Comment by jkotas** [comment]
Regression introduced by #74363 .  cc @AaronRobinsonMSFT

File: src/coreclr/debug/ee/funceval.cpp [hunk 1]
```diff
@@ -3483,7 +3483,7 @@ static void GCProtectArgsAndDoNormalFuncEval(DebuggerEval *pDE,
         ThrowHR(COR_E_OVERFLOW);
     }
     OBJECTREF * pObjectRefArray = (OBJECTREF*)_alloca(cbAllocSize);
-    *pObjectRefArray = {};
+    memset(pObjectRefArray, 0, cbAllocSize);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/ee/funceval.cpp:3486]:
```suggestion
    void* memObjRefArray = _alloca(cbAllocSize);
    memset(memObjRefArray, 0, cbAllocSize);
    OBJECTREF* pObjectRefArray = (OBJECTREF*)memObjRefArray ;
```

This is invoking undefined behavior in DEBUG/CHECKED builds. The one above also has an issue due to its fields having ctors. In DEBUG/CHECKED the `OBJECTREF` is a struct with a ctor and calling `memset` on a type with a ctor is undefined. I'm a bit surprised the `{}` wasn't zero initializing these types. Are we sure this was a case where memory wasn't zero'd out?

A simple "fix" is to `memset` the memory first and then cast.

File: src/coreclr/debug/ee/funceval.cpp [hunk 2]
```diff
@@ -3301,7 +3301,7 @@ static void DoNormalFuncEval( DebuggerEval *pDE,
         ThrowHR(COR_E_OVERFLOW);
     }
     FuncEvalArgInfo * pFEArgInfo = (FuncEvalArgInfo *)_alloca(cbAllocSize);
-    *pFEArgInfo = {};
+    memset(pFEArgInfo, 0, cbAllocSize);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/ee/funceval.cpp:3304]:
```suggestion
    void* memArgInfoArray = _alloca(cbAllocSize);
    memset(memArgInfoArray, 0, cbAllocSize);
    FuncEvalArgInfo* pFEArgInfo = (FuncEvalArgInfo*)memArgInfoArray;
```

File: src/coreclr/debug/ee/funceval.cpp [hunk 3]
```diff
@@ -3483,7 +3483,7 @@ static void GCProtectArgsAndDoNormalFuncEval(DebuggerEval *pDE,
         ThrowHR(COR_E_OVERFLOW);
     }
     OBJECTREF * pObjectRefArray = (OBJECTREF*)_alloca(cbAllocSize);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/ee/funceval.cpp:3485]:
```suggestion
```

File: src/coreclr/debug/ee/funceval.cpp [hunk 4]
```diff
@@ -3301,7 +3301,7 @@ static void DoNormalFuncEval( DebuggerEval *pDE,
         ThrowHR(COR_E_OVERFLOW);
     }
     FuncEvalArgInfo * pFEArgInfo = (FuncEvalArgInfo *)_alloca(cbAllocSize);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/ee/funceval.cpp:3303]:
```suggestion
```

File: src/coreclr/debug/ee/funceval.cpp (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/debug/ee/funceval.cpp:3486]: > I'm a bit surprised the {} wasn't zero initializing these types. Are we sure this was a case where memory wasn't zero'd out?

Ah. I missed that this logic is an array of X. This was clearly wrong, apologies.

**Review by AaronRobinsonMSFT — APPROVED** [review]
Assuming I didn't reintroduce undefined behavior into the build, this looks good. Sorry again for the regression.

**Review by tommcdon — APPROVED** [review]
LGTM, thanks!

**Pull request merged**